sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -210,8 +206,15 @@ object Cast {
     case _ => false  // overflow
   }
 
+  /**
+   * Returns `true` iff it should change the nullability of this type in map type,
+   * array type, expressions such as cast, etc.
+   *
+   * Note that you should take the nullability context into account.
+   * For example, it should not force to nullable type when null type in array
+   * type is non-nullable which means an empty array of null type.
+   */
```

Comment:
When reading this comment, I think it means `forceNullable(NullType, _)` can be true or false, depending on the context?

Reply:
But we use it like:

```
abstract class CastBase ... {
  override def nullable: Boolean = Cast.forceNullable(child.dataType, dataType) || child.nullable
  ...
}
```

Will we change `nullable` for something based on `CastBase`?

Reply:
Yes, it will also change nullability based on child's nullability. Previously, `ArrayType` of non-nullable `NullType` makes a cast `ArrayType` of nullable type always. After this change, such cast can be non-nullable based on child's nullability.

```scala
import org.apache.spark.sql.types._
import org.apache.spark.sql.functions._
spark.range(1).select(array().cast(ArrayType(IntegerType, false))).printSchema()
```

Non-nullable `NullType` should always be empty, an empty array, an empty map or an empty struct. So, it wouldn't affect the nullability in the downstream.

Reply:
How about we define `Cast.nullable` as `child.nullable || Cast.forceNullable(child.dataType, dataType)`?

Then we can say that: when calling this method, the `from` type is not nullable. If `from == NullType`, it means empty array or map, and it does not force the result to be nullable.

Reply:
Sure.